[sw/crypto] Clear HMAC ctx struct

h-filali opened the pull request:
This PR depends on #26961.

This PR adds functionality for wiping the HMAC ctx struct using randomness for the sensitive struct members.

Part of #26941.

nasahlpa commented on sw/device/lib/crypto/drivers/hmac.c:
```diff
@@ -268,6 +271,18 @@ static void msg_fifo_write(const uint8_t *message, size_t message_len) {
+ */
+static void wipe_buffer(uint32_t *buf, uint32_t buf_len) {
+  for (int i = 0; i < buf_len; ++i) {
+    buf[i] = ibex_rnd_uint32();
```
What happens if the random data inside Ibex never gets ready, e.g., when the EDN is turned off? I guess then we would loop forever in the `ibex_rnd_uint32()` function?

nasahlpa commented on sw/device/lib/crypto/impl/hash.c:
```diff
@@ -117,6 +118,24 @@ static status_t check_digest_len(otcrypto_hash_digest_t digest) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  for (int i = 0; i < kOtcryptoHashCtxStructWords; ++i) {
```
I would use `size_t` whenever possible in for loops.

nasahlpa reviewed (approved):
LGTM, thanks Hakim!

BTW, would you mind updating your mail in the git commit to your work mail?

h-filali commented on sw/device/lib/crypto/drivers/hmac.c:
```diff
@@ -268,6 +271,18 @@ static void msg_fifo_write(const uint8_t *message, size_t message_len) {
+ * @param buf The incoming buffer to be wiped with random data.
+ * @param buf_len The length of `buf` in words.
+ */
+static void wipe_buffer(uint32_t *buf, uint32_t buf_len) {
```
Sure I can use the hardened_memshred function. However, the entropy complex check can't be done in there since it doesn't have a return value. I suggest we forget the check in this PR and I do a follow up PR where I change ibex_rnd_uint to return the cycle count whenever the entropy complex is not set up properly. What do you think?

h-filali commented on sw/device/lib/crypto/drivers/hmac.c:
```diff
@@ -531,6 +581,5 @@ status_t hmac(const hmac_mode_t hmac_mode, const uint32_t *key,
   // Clean up HMAC HWIP so it can be reused by other driver calls.
   hmac_hwip_clear();
 
-  // TODO(#23191): Destroy sensitive values in the ctx object.
```
There is no `ctx` obj here. I could wipe the key and the message though. So either we leave the wiping of the input key and data to the caller or wipe them in this function.

h-filali commented:
@jadephilipoom @nasahlpa I addressed all your comments. This PR now uses `memshred` and I added randomness to the function.

jadephilipoom commented on sw/device/lib/crypto/drivers/hmac.c:
```diff
@@ -129,8 +131,10 @@ static void hmac_hwip_clear(void) {
+  // Wait until random data in ibex is ready and wipe the internal HMAC HWIP
+  // state.
+  abs_mmio_write32(kHmacBaseAddr + HMAC_WIPE_SECRET_REG_OFFSET,
+                   ibex_rnd_uint32());
```
Again, probably should have an entropy complex check before calling this (if there's a reason to believe that the entropy complex is always checked before calling this internal function, then the function documentation should reflect that as a precondition).

h-filali replied:
The function returns the cycle count if the entropy complex isn't running.

jadephilipoom commented on sw/device/lib/crypto/impl/hash.c:
```diff
@@ -117,6 +118,24 @@ static status_t check_digest_len(otcrypto_hash_digest_t digest) {
+ *
+ * @param[out] ctx Initialized context object.
+ */
+static otcrypto_status_t hash_context_wipe(otcrypto_hash_context_t *const ctx) {
```
Sorry, this comment wasn't clear -- I mean I think the function should be replaced with `hardened_memshred`. The null check isn't needed, since it's an internal function and I expect the cryptolib will always have checked if that pointer is null before getting to the point of wiping the context. So this function isn't necessary; you can just call `hardened_memshred` inline. This is actually an important distinction for code size; the `HARDENED_TRY` macro introduces several instructions and kind of blocks the compiler from just inlining the function call.

jadephilipoom commented on sw/device/lib/crypto/drivers/hmac.c:
```diff
@@ -338,12 +342,44 @@ static status_t cfg_derive(hmac_mode_t hmac_mode, uint32_t *cfg_reg,
+ * @return Result of the operation.
+ */
+static status_t hmac_context_wipe(hmac_ctx_t *ctx) {
+  if (ctx == NULL) {
+    return OTCRYPTO_BAD_ARGS;
+  }
```
You probably don't need this check, since this is an internal function. I'd double-check the call sites to be sure, but I expect that the library will generally have done something with `ctx` before wiping it. Usually in cryptolib the `NULL` checks are only for the top-level API functions, not internal ones.

h-filali replied:
Okay will do, thanks!

h-filali commented on sw/device/lib/base/hardened_memory.c:
```diff
@@ -106,7 +103,7 @@ void hardened_memshred(uint32_t *dest, size_t word_len) {
 
     // Write a freshly-generated random word to `*data`.
-    write_32(hardened_memshred_random_word(), data);
+    write_32(ibex_rnd_uint32(), data);
```
It doesn't need to. If EDN is not initialized it will just return the `mcycle_read()` value as a weak form of randomness.

jadephilipoom commented on sw/device/lib/base/ibex.c:
```diff
@@ -26,12 +28,26 @@ status_t ibex_wait_rnd_valid(void) {
-  return OK_STATUS();
+uint32_t ibex_rnd_uint32(void) {
+  uint32_t rnd_data;
+  status_t status = entropy_complex_check();
```
I think this is new since I last reviewed -- this is a dependency of the base library on a cryptolib driver. Dependencies should not go in this direction, as we've discussed recently on other PRs. The cryptolib can depend on the base library, but the base library should never ever depend on anything in `lib/crypto`. The reason for this is that the cryptolib drivers are meant very specifically for cryptolib's needs, and should not be general-purpose. For example, this check is not just "is the entropy complex awake", it's "is the entropy complex in the exact state that cryptolib expects", and I think that's not actually appropriate here. For example, ROM_EXT code might depend on this library and have EDN up but not in the cryptolib's exact expected configuration. The other concern is that this check is not exactly lightweight; doing it literally every time you pull 32 bits from RND would make reading from RND quite slow.

h-filali replied:
I'll rebase to your recent ibex change and then use the entropy_complex_check() in the respective calling functions.

jadephilipoom commented on sw/device/lib/base/ibex.c:
```diff
@@ -26,12 +28,26 @@ status_t ibex_wait_rnd_valid(void) {
+  uint32_t rnd_data;
+  status_t status = entropy_complex_check();
+
+  if (status_ok(status)) {
+    ibex_wait_rnd_valid();
+    ibex_rnd_data_read(&rnd_data);
+  } else {
+    rnd_data = (uint32_t)ibex_mcycle_read();
+  }
```
I think this is very dangerous. A single interface that might return real randomness but also might return completely non-random values (cycle counts are _very_ weak randomness, especially because the high bits are likely to repeat themselves for long periods of time) leaves a lot of opportunity for callers to expect good randomness and get something that offers no protection. For example, a caller might reasonably call `ibex_rnd_uint32()` and then use the low 16 bits for one thing and the high 16 bits for some other smaller-than-32-bit value, without realizing the high bits will be the same for 2^16 cycles at a time. Poor-quality randomness should be returned in a separate interface.

jadephilipoom reviewed (commented):
I've seen now that this PR significantly changes the behavior of the Ibex driver, which seems to be a new change since I last viewed. As best practice, I think it would make sense to separate those changes into a different PR, since it's a very different category of changes from the rest of this PR.

h-filali commented:
@nasahlpa @jadephilipoom I rebased to #27145 and removed the overlapping parts. I addressed all the requested changes. Do you think we can merge this whenever #27145 is merged?

nasahlpa commented on sw/device/lib/crypto/drivers/hmac.c:
```diff
@@ -150,14 +153,17 @@ static status_t hmac_idle_wait(void) {
-  // TODO(#23191): Use a random value from EDN to wipe.
-  abs_mmio_write32(kHmacBaseAddr + HMAC_WIPE_SECRET_REG_OFFSET, UINT32_MAX);
+  // Use a random value from EDN to wipe HMAC.
+  HARDENED_TRY(entropy_complex_check());
```
It seems that, for some functions, we check multiple times whether the entropy complex is enabled or not.

Would it make sense to move these checks at the beginning of the functions, e.g., in `hmac_final()` and remove it in `clear() `and `hmac_context_wipe()?`

h-filali replied:
Thanks Pascal, that's a good point. I'll move the check out of `clear()` and `hmac_context_wipe()`. Especially since they are static functions and we have a good overview from where the functions are called.

nasahlpa reviewed (approved):
LGTM with one question.
